### DISCUSSION TITLE
Autoformat for Neorg indentation?

### DISCUSSION THREAD

**Discussion by prmadev** [nvim-neorg/neorg#298] [General]
Hello! 
I just wanted to ask, is autoformatting, in the scope of this project? I was thinking mostly of indentation which after working for a while with neorg seems like the most annoying formatting part (e.g. when pasting a content from unindented project, they end up not conforming to the autoindentation of neorg).
If not, I would appreciate it if you know of any workarounds for indenting unindented paragraphs, instead of manually doing so. 
Thank you very much for this lovely piece of work! I grew to admire the syntax and the beautiful modular system over the past few days!

**Comment by vhyrro** [comment]
Hi! First of all thanks so much for the kind words.

I believe I should give a detailed answer here so here we go

### Formatting vs Indentation
Formatting is usually done by a third party tool or binary that runs on your file and changes its layout to something you'd desire. Indentation is the editor doing the thinking as you type so you simply don't need to think about what your document will look like, you can focus on the writing instead. I don't really know whether by "autoformat" you're asking for a formatter or for an indentation engine, but I'll assume you're talking about a native indentation engine here :P

### Shouldn't indentation be something we should be prioritizing?
Yes, but it's a lot more complex than that. Neorg has incredibly primitive indentation as you can probably tell, but this is because of how indentation works by nature. You give neovim a callback, neovim executes this callback whenever it feels like it should, and you return an integer that specifies the amount of spaces something should be indented.

Thing is we want something more than that - what about realtime indentation (one that happens as you type)? What about contextual indentation (depending on the items above or below another item we indent it differently)? Then there's single-line indents and ranged indents which have to be accounted for and given different logic as well.

---
As you can imagine, there's a lot to think about when doing indents, but it doesn't mean we won't try!
We're slowly but surely going to start working on a better indentation engine for Neorg in this PR: https://github.com/nvim-neorg/neorg/pull/291. By the end of it we hope to have an intelligent engine that will not get in your way and make writing comfy while being configurable every step of the way (since Neorg's format was designed not to care about indentation there's a lot to tweak to your personal writing style!).

Ideally we'd love to structure our code in such a way that we will be able to directly PR it with some tweaks to `nvim-treesitter` so our engine can be reapplied to actual programming languages that may have indentation issues as well. If all goes as intended we'll not only have better indents in Neorg, but also better indents overall.

Hope that explains your question! Ask away if you don't understand something I said :)

**Comment by erkanvatan** [comment]
Hello,
I came up with a hacky solution to this. Since Neorg autoindents the line on escape, I just made a keybind which jumps to the beginning of the document, repeats insert-escape until the end of the document and then jumps back to where you were in the document.
```vim
"   autoindent file in Neorg
nnoremap <silent> <localleader>i :normal! gg <CR> :%normal! i<Esc><C-o>
```

**Reply by erkanvatan** [reply]
I just learned that Neorg formats current line when you use the `=` operator in Neovim. You can mix this with Vim motions. So, `gg` to get the start of the file, `=` to indent, `G` to the end of the file, `gg=G` formats the file. Following keybind uses this and then jumps back to your previous place in the file.

```vim
"   autoindent file in Neorg
nnoremap <silent> <localleader>i gg=G<C-o>
```
